src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs

```diff
@@ -249,5 +249,37 @@ public static async Task IdentifyEncryptedEntries(string zipFile)
 
             Assert.Equal(expectedEntries, entriesEncrypted);
         }
+
+        [Fact]
+        public static async Task EnsureDisposeIsCalledOnlyOnce()
+        {
+            var disposeCallCountingStream = new DisposeCallCountingStream();
+            using (var tempStream = await StreamHelpers.CreateTempCopyStream(zfile("small.zip")))
+            {
+                tempStream.CopyTo(disposeCallCountingStream);
+            }
+
+            using (ZipArchive archive = new ZipArchive(disposeCallCountingStream, ZipArchiveMode.Read))
+            {
+                // Iterate through entries to ensure read of zip file
+                foreach (ZipArchiveEntry entry in archive.Entries)
+                {
+                    Assert.False(entry.IsEncrypted);
+                }
+            }
+
+            Assert.Equal(1, disposeCallCountingStream.NumberOfDisposeCalls);
+        }
+
+        private class DisposeCallCountingStream : MemoryStream
+        {
+            public int NumberOfDisposeCalls { get; private set; }
+
+            protected override void Dispose(bool disposing)
+            {
+                NumberOfDisposeCalls++;
+                base.Dispose(disposing);
+            }
+        }
```

Comment:
Do we have tests that validate the underlying Stream is closed when the ZipArchive is disposed according to leaveOpen?

Reply:
I have augmented the test to be a theory. It passes the leaveOpen parameter now explicitly and tests for the expected outcome. Thanks for your suggestion @stephentoub